janus/lib/models/user.rb

```diff
@@ -115,4 +115,23 @@ def has_role?(project,*roles)
         permission.project&.project_name == project
     end
   end
+  
+  def set_guest_permissions!
+    # Per project, if agreed, add guest access when no access already; remove guest access if !agreed
+    CcAgreement.where(user_email: email).all.group_by(&:project_name).each do |project_name, cc_agreements|
+      most_recent = cc_agreements.sort_by(&:updated_at).last
```

Comment:
One question: Is explicitly checking the 'update_at' necessary?

If we will always create a new entry whenever a user interacts with the CoC-ui, then 1) we could simplify to cc_agreements.last as this would (I think) always be the most recent, and 2) we also wouldn't need an 'updated_at' field as it would be redundant with 'created_at'.

Reply:
Good questions.

In general, I do believe that `created_at` should be redundant with `updated_at`.  This table absolutely should be append only immutable.  That said, there is one plausible way in which `updated_at` might be meaningful: if we ever extend the table and need to provide backfill.  It's, unlikely but not impossible.  In that case, `updated_at` is more of a 'debug' column that helps us potentially understand backfill state.

Ordering by id does benefit from immutability in ways that relying on `updated_at` or `created_at` does not, but to be clear, **there is not a guarantee of monotonic increasing values** from id.  99% of the time the id will increment, but it's important to remember how auto increments are actually, implemented.  They are state that can be manipulated (you can set the autoincrement with SQL commands directly) by bugs, and furthermore, they are by default backed by 32 bit integers which can overflow (in which case the ids will actually return to the lowest unused value in the table, in case of gaps, which could occur due to failed transactions or deletes).

The most likely case of id overflowing is a bug: something accidentally inserts or attempts to insert a huge amount of bad data and forces the id to overflow.  It, does happen.  But, not often.

My intuition is that `created_at`, `id` is the best ordering here, leaving `updated_at` as merely debug information in the case of necessary future backfill.

Reply:
Swapped to sorting by `created_at`.  Assuming it would resort to `id` next if there ends up being a `created_at` tie, seems like that gets the intended behavior